### PR TITLE
Convert config.json to dockercfg format as needed (#55)

### PR DESCRIPTION
**Pull request by jonathanstallings**
* Allow l0-setup to read docker config from `config.json` or `dockercfg`
* Write `dockercfg` format to instance directory
* Allow `--docker-config-dir` flag to specify directory to read config from
* Attempt to read from global docker config dir if not found at given config directory
* Fall back to writing "empty" dockercfg file (`{}`) if needed
* Preserve `--force` flag, which disables alert on `empty` dockercfg

Would love any feedback you have, especially regarding preferences for where to organize the code additions such as `/setup/context/dockerauth.go`.

### CODE REVIEW

**Comment by jonathanstallings** [comment]
Notes:
Some guidance found by reference to code within the docker repository around time when `config.json` support was added.
See: https://github.com/docker/docker/blob/18c9b6c6455f116ae59cde8544413b3d7d294a5e/registry/auth.go

Sections of `dockerauth.go` may be simplified and attribution to reference code may be added as needed.

File: setup/context/dockerauth.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	OLD_CONFIGFILE = "dockercfg"
+)
+
+// AuthConfig contains authorization information for connecting to a Registry.
+type AuthConfig struct {
+	Username string `json:"username,omitempty"`
+	Password string `json:"password,omitempty"`
+	Auth     string `json:"auth,omitempty"`
+
+	// Email is an optional value associated with the username.
```
Comment (zpatrick) [inline setup/context/dockerauth.go]: we don't need these comments

File: setup/context/dockerauth.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+type DockerConfigFile struct {
+	AuthConfigs map[string]AuthConfig `json:"auths"`
+	HTTPHeaders map[string]string     `json:"HttpHeaders,omitempty"`
+	filename    string                // Note: not serialized - for internal use only
+	Empty       bool                  // Note: not serialized - for internal use only
+}
+
+// LoadDockerConfig loads docker config files of config.json or dockercfg formats.
+func LoadDockerConfig(configDir string) (*DockerConfigFile, error) {
+	if configDir == "" {
```
Comment (zpatrick) [inline setup/context/dockerauth.go]: remove param checking/validation - that's a python-esq pattern you typically don't do w/ go

File: setup/context/dockerauth.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	RegistryToken string `json:"registrytoken,omitempty"`
+}
+
+// DockerConfigFile contains `config.json` file info.
+// `dockercfg` files use just the `AuthConfigs` field.
+type DockerConfigFile struct {
+	AuthConfigs map[string]AuthConfig `json:"auths"`
+	HTTPHeaders map[string]string     `json:"HttpHeaders,omitempty"`
+	filename    string                // Note: not serialized - for internal use only
+	Empty       bool                  // Note: not serialized - for internal use only
```
Comment (zpatrick) [inline setup/context/dockerauth.go]: don't need to comment on internal use, it's a given from lowercase fields

Comment (zpatrick) [inline setup/context/dockerauth.go]: if `Empty` is private why is it capitalized?

File: setup/context/dockerauth.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	configFile := DockerConfigFile{
+		AuthConfigs: make(map[string]AuthConfig),
+		filename:    filepath.Join(configDir, CONFIGFILE),
+	}
+
+	// Try config.json file first
+	if _, err := os.Stat(configFile.filename); err == nil {
+		file, err := os.Open(configFile.filename)
+		if err != nil {
+			return &configFile, err
```
Comment (zpatrick) [inline setup/context/dockerauth.go]: `return nil, err`

File: setup/context/dockerauth.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+		return nil, fmt.Errorf("No configdir supplied.")
+	}
+
+	configFile := DockerConfigFile{
+		AuthConfigs: make(map[string]AuthConfig),
+		filename:    filepath.Join(configDir, CONFIGFILE),
+	}
+
+	// Try config.json file first
+	if _, err := os.Stat(configFile.filename); err == nil {
```
Comment (zpatrick) [inline setup/context/dockerauth.go]: always check `err != nil`, and don't created nested blocks if you can help it. This should be:
```
if _, err := os.Stat(configFile.filename); err != nil {
    return nil, err
}

file, err := os.Open(configFile.filename)
...

```

File: setup/context/dockerauth.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	data, err := json.MarshalIndent(configFile.AuthConfigs, "", "\t")
+	if err != nil {
+		return err
+	}
+
+	if err := os.MkdirAll(filepath.Dir(saveDir), 0600); err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(filepath.Join(saveDir, OLD_CONFIGFILE), data, 0600)
```
Comment (zpatrick) [inline setup/context/dockerauth.go]: [nit] `if err := iotuil.WriteFile(...); err != nil{ ... }`

File: setup/main.go [hunk 7]
```diff
@@ -76,15 +76,15 @@ func main() {
 	app.Command("apply", "Create/Update a Layer0", func(cmd *cli.Cmd) {
 		flags := loadFlags(cmd, []string{"access_key", "secret_key", "region"})
 
-		dockercfg := cmd.StringOpt("dockercfg", "", "Path to valid dockercfg file")
+		dockerCfgDir := cmd.StringOpt("docker-config-dir", "", "Path to directory containing valid dockercfg or config.json file")
```
Comment (zpatrick) [inline setup/main.go]: I think this should stay as a path to a file:
```
dockercfg := cmd.StringOpt("dockercfg", "", "Path to valid dockercfg or config.json file")
```

File: setup/context/commands.go [hunk 8]
```diff
@@ -63,17 +65,9 @@ func Apply(c *Context, force bool, dockercfg string) error {
...
-		if err := CopyFile(dockercfg, instancePath); err != nil {
-			return fmt.Errorf("Failed to copy dockercfg: %v.", err)
-		}
-	}
-
-	// validate dockercfg; if file does not exist, write empty dockercfg file
-	if err := checkDockercfg(c, force); err != nil {
+	// Check for docker config file (config.json or dockercfg) and copy it to
+	// instance directory in `dockercfg` format.
+	if err = checkDockercfg(c, dockerCfgDir, force); err != nil {
```
Comment (zpatrick) [inline setup/context/commands.go]: [nit] don't need to propagate `force` behavior across too many function if we can avoid it:
```
if !force {
    if err := checkDockercfg(c, dockerCfgDir); err != nil{
        return err
    }
}
```

File: setup/context/commands.go [hunk 9]
```diff
@@ -408,29 +402,55 @@ func Endpoint(c *Context, syntax string, insecure, dev, quiet bool) error {
...
 	args = append([]string{"plan"}, args...)
 	return Terraform(c, args)
 }
 
-func checkDockercfg(c *Context, force bool) error {
-	path := fmt.Sprintf("%s/dockercfg", c.InstanceDir)
+// checkDockercfg looks for docker config file (config.json or dockercfg) from
+// given directory and saves in `dockercfg` format in instance directory.
+func checkDockercfg(c *Context, dockerCfgDir string, force bool) error {
+	if dockerCfgDir == "" {
```
Comment (zpatrick) [inline setup/context/commands.go]: we don't use default params like this in go, this logic should exist in the caller

File: setup/context/commands.go [hunk 10]
```diff
@@ -408,29 +402,55 @@ func Endpoint(c *Context, syntax string, insecure, dev, quiet bool) error {
...
+		dockerCfgDir = c.InstanceDir
+	} else {
+		fmt.Printf("Checking for docker config file at: `%s`\n", dockerCfgDir)
+	}
+
+	// Load docker config file from given dockerCfgDir
+	configFile, err := LoadDockerConfig(dockerCfgDir)
```
Comment (zpatrick) [inline setup/context/commands.go]: For the sake of keeping things simple, I don't think we need this "discovery" behavior. 
Specifying a file path via `--dockercfg=path/to/file` is all we need to implement.

File: setup/context/dockerauth.go (hunk 1 above)
Comment (jonathanstallings) [inline setup/context/dockerauth.go]: Removed.

File: setup/context/dockerauth.go (hunk 3 above)
Comment (jonathanstallings) [inline setup/context/dockerauth.go]: Comments removed.

Comment (jonathanstallings) [inline setup/context/dockerauth.go]: Good catch. Changed.

File: setup/context/dockerauth.go (hunk 6 above)
Comment (jonathanstallings) [inline setup/context/dockerauth.go]: 👍

File: setup/context/commands.go (hunk 8 above)
Comment (jonathanstallings) [inline setup/context/commands.go]: I was under the impression that the `--force` flag was only intended to skip prompting on missing `dockercfg` file. However, the "empty" `dockercfg` file still needs to be created if missing, correct?

This suggestion would skip creating that file if missing.

File: setup/context/commands.go (hunk 9 above)
Comment (jonathanstallings) [inline setup/context/commands.go]: I have seen this pattern in go source code, and I was modeling from it. Definitely will change if you prefer it differently, as I want to set this up the proper way. Thanks for the tip.

Moving default logic to caller.

**Comment by jonathanstallings** [comment]
I'd be very happy to pair on this @zpatrick. Thanks for the offer!

File: setup/context/dockerauth.go (hunk 2 above)
Comment (jonathanstallings) [inline setup/context/dockerauth.go]: Removed.

File: setup/context/commands.go (hunk 10 above)
Comment (jonathanstallings) [inline setup/context/commands.go]: Removed discovery behavior and reverted to specifying file path explicitly.

File: setup/context/commands.go (hunk 8 above)
Comment (diemonster) [inline setup/context/commands.go]: @jonathanstallings I think zpatrick is just saying that passing `force` as a param/behavior around to other functions, even if they are helpers, is not ideal.

File: setup/main.go (hunk 7 above)
Comment (jonathanstallings) [inline setup/main.go]: Changed to file path as requested.

File: setup/context/dockerauth.go (hunk 5 above)
Comment (jonathanstallings) [inline setup/context/dockerauth.go]: Changed to `err != nil` pattern. Thanks!

File: setup/context/dockerauth.go (hunk 4 above)
Comment (jonathanstallings) [inline setup/context/dockerauth.go]: Using the the existing pattern facilitates saving an "empty" config. However, I can certainly change this.

**Comment by diemonster** [comment]
@zpatrick @jonathanstallings 

Re: ripping out the "discovery behavior". I'm a bit torn on this, but it's a workflow issue more general to all of our CLI UX. 

We consistently expect the user to know about specific flags to circumvent default behavior, but in some cases (like this) our default behavior is not the most common use case. Beyond that, we push the burden of trial and error onto the user and force them to muck through our tools until they get the perfect combination of arguments. This is frustrating for many users, and I'm worried it will eventually bite us if Layer0 is adopted beyond our close confines. 

Zpatrick's suggestion is in line with how the rest of our CLI UX works, so in that sense, I agree w his comment _right now_. But thinking from a bigger picture, we need to consider adding actual logic in our CLI tools to prevent users from giving up too quickly. `l0-setup` for all of it's deficiencies, is the introduction to our entire toolset for new people, and we need to put a friendlier foot forward.

**Comment by zpatrick** [comment]
closing for different branch

**Pull request closed without merging**